### PR TITLE
custom_punctuation macro

### PR DESCRIPTION
**Pull request by CAD97**
Follow up to https://github.com/dtolnay/syn/pull/503 (#568 is next).

Provide a `custom_punctuation!` analog to `custom_keyword!`.

Getting back into the swing of working with syn, and have want for these patches again. So let's finish them up properly!

### CODE REVIEW

**Comment by CAD97** [comment]
OK, failure is that the `Tests` job is using `custom_keyword` without the `parsing` feature, and `trait CustomToken` is gated on the `parsing` feature. I'm not sure the exact place where this should be fixed, so I'm going to wait for someone with a better knowledge of the feature matrix to suggest the proper solution.

**Comment by dtolnay** [comment]
The custom_keyword doc tests are expanding to code that does not compile, and the custom_punctuation test panics when run. I don't think it is a feature matrix issue. You can reproduce it locally with: `cargo test --all-features --doc`

File: src/custom_token.rs [hunk 1]
```diff
@@ -0,0 +1,536 @@
...
+    (;)      => ( 1usize );
+    (<<)     => ( 2usize );
+    (<<=)    => ( 3usize );
+    (>>)     => ( 2usize );
+    (>>=)    => ( 3usize );
+    (*)      => ( 1usize );
+    (-)      => ( 1usize );
+    (-=)     => ( 2usize );
+    (~)      => ( 1usize );
+    ($tt:tt) => ( compile_error!("Punctuation `{}` is not supported", stringify!($tt)) );
```
Comment (dtolnay) [inline src/custom_token.rs]: This produces:

```console
error[E0433]: failed to resolve: could not find `compile_error` in `$crate`
 --> src/main.rs:2:5
  |
2 |     syn::custom_punctuation!(LeftRightArrow, <x>);
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ could not find `compile_error` in `$crate`
```

Comment (CAD97) [reply src/custom_token.rs]: ugh local_inner_macros is weird

**Comment by CAD97** [comment]
Ah, I messed up the merge of my branch and master.

**Comment by CAD97** [comment]
<sup>I'm going to claim [it's not my fault](https://github.com/intellij-rust/intellij-rust/issues/3707) I missed this error, I actually ran tests this time</sup>

That's the doc test finally working as intended. The operator is kind of clunky in that it requires parenthesized expressions, but it gets the point across decently, I think.

File: src/custom_token.rs [hunk 2]
```diff
@@ -0,0 +1,554 @@
...
+    (;)      => ( 1usize );
+    (<<)     => ( 2usize );
+    (<<=)    => ( 3usize );
+    (>>)     => ( 2usize );
+    (>>=)    => ( 3usize );
+    (*)      => ( 1usize );
+    (-)      => ( 1usize );
+    (-=)     => ( 2usize );
+    (~)      => ( 1usize );
+    ($tt:tt) => ( my_compile_error!("Punctuation `{}` is not supported", stringify!($tt)) );
```
Comment (dtolnay) [inline src/custom_token.rs]: Now it says:

```console
error: compile_error! takes 1 argument
 --> src/main.rs:2:5
  |
2 |     syn::custom_punctuation!(LeftRightArrow, <x>);
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Comment (CAD97) [reply src/custom_token.rs]: I gave up trying to be too helpful and just gave a simple enough error message

**Comment by dtolnay** [comment]
Thanks. I went ahead and pushed my own small change to the error reporting.

#### Before:

```console
error: This punctuation is not supported
 --> src/main.rs:2:5
  |
2 |     syn::custom_punctuation!(LeftRightArrow, <x>);
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

error: This punctuation is not supported
 --> src/main.rs:2:5
  |
2 |     syn::custom_punctuation!(LeftRightArrow, <x>);
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

error: This punctuation is not supported
 --> src/main.rs:2:5
  |
2 |     syn::custom_punctuation!(LeftRightArrow, <x>);
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

error[E0308]: mismatched types
 --> src/main.rs:2:5
  |
2 |     syn::custom_punctuation!(LeftRightArrow, <x>);
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected slice, found array
  |
  = note: expected type `&[proc_macro2::Span]`
             found type `&[proc_macro2::Span; _]`
  = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

error: aborting due to 4 previous errors
```

```console
error: unexpected end of macro invocation
 --> src/main.rs:2:5
  |
2 |     syn::custom_punctuation!(Nothing, );
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing tokens in macro arguments
  |
  = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

error: unexpected end of macro invocation
 --> src/main.rs:2:5
  |
2 |     syn::custom_punctuation!(Nothing, );
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing tokens in macro arguments
  |
  = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

error: unexpected end of macro invocation
 --> src/main.rs:2:5
  |
2 |     syn::custom_punctuation!(Nothing, );
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing tokens in macro arguments
  |
  = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)

error[E0308]: mismatched types
 --> src/main.rs:2:5
  |
2 |     syn::custom_punctuation!(Nothing, );
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected slice, found array
  |
  = note: expected type `&[proc_macro2::Span]`
             found type `&[proc_macro2::Span; _]`
  = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)
```

#### After:

```console
error: no rules expected the token `x`
 --> src/main.rs:2:47
  |
2 |     syn::custom_punctuation!(LeftRightArrow, <x>);
  |                                               ^ no rules expected this token in macro call
```

```console
error: unexpected end of macro invocation
 --> src/main.rs:2:38
  |
2 |     syn::custom_punctuation!(Nothing, );
  |                                      ^ missing tokens in macro arguments
```

**Pull request merged**